web/modules/custom/dpl_related_content/dpl_related_content.module

```diff
@@ -0,0 +1,292 @@
+<?php
+
+use Drupal\Core\Database\Database;
+use Drupal\Core\Datetime\DrupalDateTime;
+use Drupal\Core\Entity\FieldableEntityInterface;
+use Drupal\datetime\Plugin\Field\FieldType\DateTimeItemInterface;
+use Drupal\node\Entity\Node;
+use Drupal\node\NodeInterface;
+use Drupal\recurring_events\Entity\EventInstance;
+
+/**
+ * Get terms from field as an array of IDs.
+ *
+ * @return array<int>
+ *   Term IDs.
+ */
+function _dpl_related_content_get_term_ids(FieldableEntityInterface $entity, string $field_name): array {
+  if (!$entity->hasField($field_name)) {
+    return [];
+  }
+
+  $terms = $entity->get($field_name)->getValue();
+  return array_column($terms, 'target_id');
+
+}
+
+/**
+ * Get matching eventinstance IDs.
+ *
+ * Allows for passing along various term IDs, that we look for in an OR group.
+ * Notice - we only look for the term values on series level, ignoring
+ * inheritance overriding in favor of a simpler codebase.
+ *
+ * @param string|null $current_uuid
+ *   A possible entity UUID, that will not get included in results.
+ * @param array<int> $tags
+ *   Tag term IDs, to look for.
+ * @param array<int> $categories
+ *   Category term IDs, to look for.
+ * @param array<int> $branches
+ *   Branch term IDs, to look for.
+ *
+ * @return array<int|string>
+ *   Matching eventinstance IDs
+ *
+ * @throws \Exception
+ */
+function _dpl_related_content_get_eventinstance_ids(?string $current_uuid, array $tags = [], array $categories = [], array $branches = []): array {
+  $date = new DrupalDateTime('today');
+  $date->setTimezone(new \DateTimezone(DateTimeItemInterface::STORAGE_TIMEZONE));
+  $formatted_date = $date->format(DateTimeItemInterface::DATETIME_STORAGE_FORMAT);
+
+  $connection = Database::getConnection();
```

Comment:
I would expect some kind of comment about why we use direct queries here instead of an entity query.

Perhaps it is because of the subquery?

Reply:
Indeed because of the subquery. I'll add a comment about it.